File: frontend-research/src/services/notificationService.js
/**
 * Notification API service
 * Wraps all notification-related backend API calls.
 */
import api from './api';

const notificationService = {
  /**
   * Fetch all notifications for a user.
   * GET /api/notifications?email=...
   */
  getAll: async (email) => {
    return await api.get(`/notifications?email=${encodeURIComponent(email)}`);
  },

  /**
   * Fetch only unread notifications (for the bell panel).
   * GET /api/notifications/unread?email=...
   */
  getUnread: async (email) => {
    return await api.get(`/notifications/unread?email=${encodeURIComponent(email)}`);
  },

  /**
   * Get the unread notification count (for the badge).
   * GET /api/notifications/unread-count?email=...
   */
  getUnreadCount: async (email) => {
    return await api.get(`/notifications/unread-count?email=${encodeURIComponent(email)}`);
  },

  /**
   * Mark a single notification as read.
   * PUT /api/notifications/{id}/read
   */
  markAsRead: async (id) => {
    return await api.put(`/notifications/${id}/read`);
  },

  /**
   * Mark all notifications for a user as read.
   * PUT /api/notifications/read-all?email=...
   */
  markAllAsRead: async (email) => {
    return await api.put(`/notifications/read-all?email=${encodeURIComponent(email)}`);
  },
};

export default notificationService;

File: backend-research/src/main/java/com/frankmoley/lil/backendresearch/service/NotificationService.java
package com.frankmoley.lil.backendresearch.service;

import com.frankmoley.lil.backendresearch.dto.NotificationDTO;

import java.util.List;

/**
 * Abstraction interface for notification business logic.
 * Demonstrates OOP Abstraction: defines a contract that concrete implementations must fulfill.
 */
public interface NotificationService {

    /**
     * Get all notifications for a user email.
     *
     * @param userEmail the user's email
     * @return list of notification DTOs, newest first
     */
    List<NotificationDTO> getNotificationsForUser(String userEmail);

    /**
     * Get only unread notifications for a user (for the bell badge/panel).
     *
     * @param userEmail the user's email
     * @return list of unread notification DTOs
     */
    List<NotificationDTO> getUnreadNotifications(String userEmail);

    /**
     * Get the count of unread notifications for a user.
     *
     * @param userEmail the user's email
     * @return number of unread notifications
     */
    long getUnreadCount(String userEmail);

    /**
     * Mark a single notification as read.
     *
     * @param notificationId the notification ID
     * @return updated notification DTO
     */
    NotificationDTO markAsRead(Long notificationId);

    /**
     * Mark all notifications for a user as read.
     *
     * @param userEmail the user's email
     */
    void markAllAsRead(String userEmail);
}
